TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java

```diff
@@ -19,75 +23,85 @@
+    protected final EmbeddedChannel httpDecoderChannel;
+    SimpleHttpRequestDecoder requestDecoder;
 
-    private DefaultHttpRequest currentHttpRequest;
 
     public LoggingHttpRequestHandler(IChannelConnectionCaptureSerializer trafficOffloader) {
         this.trafficOffloader = trafficOffloader;
-        httpDecoder = new SimpleHttpRequestDecoder();
-        currentHttpRequest = null;
+        requestDecoder = new SimpleHttpRequestDecoder(); // as a field for easier debugging
+        httpDecoderChannel = new EmbeddedChannel(
+                requestDecoder,
+                new SimpleDecodedHttpRequestHandler()
+        );
     }
 
+
+
     protected HttpCaptureSerializerUtil.HttpProcessedState
     onHttpObjectsDecoded(List<Object> parsedMsgs) throws IOException {
         parsedMsgs.stream()
                 .filter(o -> o instanceof DefaultHttpRequest)
                 .map(o -> (DefaultHttpRequest) o)
                 .findAny()
-                .ifPresent(req -> this.currentHttpRequest = req);
+                .ifPresent(req -> getHttpRequestReceiverHandler().currentRequest = req);
         return HttpCaptureSerializerUtil.addRelevantHttpMessageIndicatorEvents(trafficOffloader, parsedMsgs);
     }
 
     private HttpCaptureSerializerUtil.HttpProcessedState parseHttpMessageParts(ChannelHandlerContext ctx, ByteBuf msg) throws Exception {
         var bb = msg;
+        // Preserve ownership of this ByteBuf because the HttpRequestDecoder will take ownership of the
+        // ByteBuf, releasing it once the data has been copied into its cumulation buffer.  However,
+        // this handler still fires ByteBufs through the pipeline.  It's up to a future handler to perform
+        // the release, as it would have been if this method was implemented any other way.
+        bb.retain();
         bb.markReaderIndex();
-        // todo - move this into a pool
-        var rval = HttpCaptureSerializerUtil.HttpProcessedState.ONGOING;
-        var parsedMsgs = new ArrayList<>(4);
-        while (bb.readableBytes() > 0) {
-            var lastIdx = bb.readerIndex();
-            log.warn("Parsing from idx="+lastIdx+" readableBytes="+ bb.readableBytes());
-            httpDecoder.decode(ctx, bb, parsedMsgs);
-            rval = onHttpObjectsDecoded(parsedMsgs);
-            if (rval == HttpCaptureSerializerUtil.HttpProcessedState.FULL_MESSAGE || // got what we needed
-                    lastIdx == bb.readerIndex()) { // no progress
-                break;
-            }
-            parsedMsgs.clear();
-        }
-        log.trace("Resetting index");
+        httpDecoderChannel.writeInbound(bb);
         bb.resetReaderIndex();
-        return rval;
+        return getHttpRequestReceiverHandler().isDone ?
+                HttpCaptureSerializerUtil.HttpProcessedState.FULL_MESSAGE :
+                HttpCaptureSerializerUtil.HttpProcessedState.ONGOING;
+    }
+
+    private SimpleDecodedHttpRequestHandler getHttpRequestReceiverHandler() {
```

Comment:
nit: the naming being different from type and method name threw me off a little

Reply:
Slightly, my inclination is to use `getSimpleDecodedHttpRequestHandler` but that might not be clear to others so I'm fine either way